x-pack/plugins/fleet/server/routes/epm/handlers.ts

```diff
@@ -102,22 +106,51 @@ export const getFileHandler: RequestHandler<TypeOf<typeof GetFileRequestSchema.p
 ) => {
   try {
     const { pkgName, pkgVersion, filePath } = request.params;
-    const registryResponse = await getFile(`/package/${pkgName}/${pkgVersion}/${filePath}`);
-
-    const headersToProxy: KnownHeaders[] = ['content-type', 'cache-control'];
-    const proxiedHeaders = headersToProxy.reduce((headers, knownHeader) => {
-      const value = registryResponse.headers.get(knownHeader);
-      if (value !== null) {
-        headers[knownHeader] = value;
+    const savedObjectsClient = context.core.savedObjects.client;
+    const savedObject = await getInstallationObject({ savedObjectsClient, pkgName });
+    const pkgInstallSource = savedObject?.attributes.install_source;
+    // TODO: when package storage is available, remove installSource check and check cache and storage, remove registry call
+    if (pkgInstallSource === 'upload') {
+      const headerContentType = mime.contentType(path.extname(filePath));
+      if (!headerContentType) {
+        return response.custom({
+          body: 'file type is invalid',
+          statusCode: 400,
+        });
       }
-      return headers;
-    }, {} as ResponseHeaders);
+      const archiveFile = getArchiveEntry(`${pkgName}-${pkgVersion}/${filePath}`);
+      if (!archiveFile) {
+        return response.custom({
+          body: 'file does not exist in cache',
+          statusCode: 404,
+        });
+      }
+      const headers: ResponseHeaders = {
+        'cache-control': 'max-age=10, public',
```

Comment:
I'm not sure what the correct cache value is, but this is the shortest time we have. It's what's used for packages from snapshot Registry.

My initial reaction is to cache for a loooong time. Like, forever. At least a year. Since the a given package version is supposed to be immutable, but I don't know how that will work in the "developer working on a integration locally" case. We wouldn't want to cache the initial version so long they couldn't see their iterative changes. On the other hand, not every uploaded package is (or will be) from a developer on a short iterative cycle.

Reply:
Good points.  I am leaning toward keeping it short for now seeing it will mostly be a developer thing or a mostly temporary thing?

Reply:
Totally. We can put this in now and ask questions / change later.